### ISSUE TITLE
#gmcr generates unknown array_wget and GMEdit wrong warnings / errors

### ISSUE THREAD

**Issue by Matoneto** [YellowAfterlife/GMEdit#196]
I would like to use the coroutines feature of GMEdit and tried to run the array_foreach example. I created a script and pasted the code from the example in it:

```js
/// array_foreach(array)
#gmcr
var arr = argument0;
var len = array_length_1d(arr);
for (var i = 0; i < len; i++) {
    yield arr[i];
}
````

Then i created an object and in its create event i pasted:
```js
#event create
var myarray = [1,2,3];
var iter = array_foreach(0, myarray);
while (array_foreach(iter)) {
    show_message(iter[0]);
}
````

I placed the object in a room and tried to start it.

Gamemaker cannot find the function or script for array_wget. I tried it in GMS1.4 and GMS2 2.2.5.

Another thing is that GMEdit gives warnings about array_foreach when calling it with different argument numbers (first call with two arguments and all following calls only with one) and about it not returning anything when it is called in the while loop. It always expects one argument and no return value although it should "see" the generated code i think.

The following code is generated for the coroutine:

````js
/// array_foreach(ctx, array)
var l_ctx = argument[0];
if (!is_array(l_ctx)) {
    l_ctx = array_create(6);
    var l_argc = argument_count - 1;
    var l_args = array_create(l_argc);
    while (--l_argc >= 0) l_args[@l_argc] = argument[l_argc + 1];
    l_ctx[2/* args */] = l_args;
    return l_ctx;
}
var l_args = l_ctx[2/* args */];
while (true) switch (l_ctx[1/* label */]) {
case 0/* [L1,c1] begin */:
    l_ctx[@3/* arr */] = l_args[0];
    l_ctx[@4/* len */] = array_length_1d(l_ctx[3/* arr */]);
    l_ctx[@5/* i */] = 0;
case 1/* [L5,c1] check for */:
    if (l_ctx[5/* i */] >= l_ctx[4/* len */]) { l_ctx[@1/* label */] = 4; continue; }
    l_ctx[@0/* yield */] = array_wget(l_ctx[3/* arr */], l_ctx[5/* i */]); l_ctx[@1/* label */] = 2; return true;
case 2/* [L6,c5] post yield, [L5,c1] post for */:
    l_ctx[@5/* i */]++;
    l_ctx[@1/* label */] = 1; continue;
default/* [L5,c1] end for, [L7,c2] end */: l_ctx[@0/* yield */] = 0; return false;
}

/*//!#gmcr
/// array_foreach(array)
#gmcr
var arr = argument0;
var len = array_length_1d(arr);
for (var i = 0; i < len; i++) {
    yield arr[i];
}
//!#gmcr*/
```

**Comment by YellowAfterlife** [comment]
`array_wget` and `array_wset` were for GMS1, which didn't expose `array_get` and `array_set` yet. So you could do
```gml
#macro array_wget array_get
#macro array_wset array_set
```
However, per wiki, coroutine generator isn't very up to date in general - it's a funny idea that I did for some work in 2019, but it didn't gather much interest even back then, and now it would take several days of work to make it support new GML syntax and properly wrap code in `function <name>()` and whatnot.

**Comment by Matoneto** [comment]
Thanks for the help. Adding the macros fixed the missing function error. It now works as expected.

What about the warnings GMEdit gives me about argument counts and return value? Is it because GMEdit does not see the generated code and the warnings are based on the code as written?

I think coroutines are one of the most important missing features of GM so I am surprised this didnt get a lot of attention. I have a use case where i want to make a bullet hell game and with coroutines i want to be able to give bullets their special behaviour (like in [Touhou](https://en.wikipedia.org/wiki/Touhou_Project) games). So i expect there to be many coroutines at the same time and performance wise i think this approach should be much better compared to the alternative 2.3+ macro way mentioned on the coroutine wiki page.

Your approach only uses one array per coroutine for the state and switch cases for the continuation. The macro approach creates a struct for the state and many functions inside for the continuation. I'd prefer your approach.

**Comment by Matoneto** [comment]
I have a few questions.

1. What about the warnings GMEdit gives me about argument counts and return value? Is it because GMEdit does not see the generated code and the warnings are based on the code as written?
2. Is the coroutine feature production ready for my use case described above or is it considered deprecated?

**Comment by YellowAfterlife** [comment]
1. I think so, but also coroutines are much older than the linter so there might be caveats.
2. I'd suggest to wait until I fix these up - perhaps I could also add an option to generate a function that returns a single coroutine method instead, which would make for a slightly nicer
  ```
  #event create
  var myarray = [1,2,3];
  var iter = array_foreach(myarray);
  while (iter()) {
      show_message(iter.result);
  }
  ```
  or something alone those lines

**Comment by Matoneto** [comment]
This option sounds good for 2.3+ users. Do you think performance will be good enough for a few hundered of these at the same time?

Thanks again for this coroutine feature and for improving it.

**Comment by YellowAfterlife** [comment]
Creating a function when a coroutine is created isn't too expensive.
 
Also, re-reading your original description, if you want to use this in 1.4 or 2.2.5, go ahead - the feature won't change much as far as your uses go.

**Comment by Matoneto** [comment]
For these older versions of GM the only problem that remains are the warnings GMEdit shows because of argument numbers and return types I mentioned.

**Comment by Matoneto** [comment]
I did some coroutine performance tests. I spawned coroutines (10 every step up to 5000 max) that each do some calculations every frame:

- GM 2023.6 with MACRO based coroutines: Drops below 60fps at 2200 and maintains 27fps at 5000.
- GMS 2.2.5 with GMEdit generated coroutine code: Drops below 60fps at 3500 and maintains 35fps at 5000.
- GM 2023.6 with GMEdit generated coroutine code manually wrapped in a function: No drop below 60fps, maintains 130fps at 5000.

As I thought the MACRO based approach is the slowest but I was really surprised that the new GM Version runs so much faster than 2.2.5. It seems they improved performance quite a bit and the future runtime with native compilation will probably improve this even more.

So conclusion for me is that your coroutine feature offers much better performance and code readability (before transformation) compared to the MACRO library. When it comes to debugging coroutine code both can give a hard time though.
